libs/core/langchain_core/indexing/schema_02.py

```diff
@@ -0,0 +1,235 @@
+    succeeded: NotRequired[Sequence[str]]
+    failed: NotRequired[Sequence[str]]
+
+
+class Sort(TypedDict):
+    """A sort object."""
+
+    field: str
+    ascending: NotRequired[bool]  # Assume asc=True
+
+
+C = TypeVar("C")
+
+
+class BaseIndex(Generic[T]):
+    """An index represent a collection of items that can be queried.
+
+    The index is responsible for:
+
+    1. Storing the content
+    2. Allowing retrieval of the content by id
+    3. Supporting queries against the metadata associated with the content
+
+    The index is **NOT** responsible for:
+
+    1. Support search queries against the content.
+
+    The types of supported queries seems very dependent on the actual
+    """
+
+    @abc.abstractmethod
+    def upsert(self, data: Iterable[T], /, **kwargs: Any) -> Iterable[UpsertResponse]:
+        """Upsert a stream of data by id."""
+
+    @abc.abstractmethod
+    def get_by_ids(
+        self,
+        ids: Sequence[str],
+        /,
+    ) -> List[T]:
+        """Get items by id."""
+
+    @abc.abstractmethod
+    def delete_by_ids(
+        self,
+        ids: Sequence[str],
+        /,
+        **kwargs: Any,
+    ) -> DeleteResponse:
+        """Delete an item by id."""
+
+    # Delete and get are part of the READ/WRITE interface.
+    # They do not take advantage of indexes on the content.
+    # However, all the indexers ARE assumed to have the capability to index
+    # on metadata if they implement the delete_by_query and get_by_query methods.
+    @abc.abstractmethod
+    def get_by_query(
+        self,
+        *,
+        filter: Optional[Union[List[Dict[str, Any], Dict[str, Any]]]] = None,
+        limit: Optional[int] = None,
+        offset: Optional[int] = None,
+        sort: Optional[Union[Sort, List[Sort]]] = None,
+        **kwargs: Any,
+    ) -> Iterable[T]:
+        """Get items by query."""
+
+    @abc.abstractmethod
+    def delete_by_query(
+        self,
+        *,
+        filter: Optional[Union[List[Dict[str, Any], Dict[str, Any]]]] = None,
+        limit: Optional[int] = None,
+        offset: Optional[int] = None,
+        sort: Optional[Union[Sort, List[Sort]]] = None,
+        **kwargs: Any,
+    ) -> Iterable[DeleteResponse]:
+        """Delete items by query."""
+        # Careful with halloween problem
+
+
+class Hit(Generic[T], TypedDict):
+    # Should subclass from content?
+    id: str
+    score: float
+    # Optional source
+    source: Optional[T]
+    metadata: Dict[str, Any]
+
+
+class QueryResponse(TypedDict):
+    """A retrieval result."""
+
+    # Free form metadata for vectorstore providers
+    metadata: Dict[str, Any]
+    hits: List[Hit]
+
+
+class UpsertData(TypedDict):
+    """A structured representation of a data item to upsert.
+
+    This allows providing additional information such as the vector.
+    """
+
+    id: str  # An override of the ID for the document
+    document: Document
+    vector: List[float]
+
+
+class VectorStore(BaseIndex[Document]):
```

Comment:
Vectorstore implementation over Document, but can look differently for a vectorstore over images (i.e., different query)